File: app/src/test/java/com/example/mike/mp3player/service/library/content/retriever/SongsFromFolderRetrieverTest.java
package com.example.mike.mp3player.service.library.content.retriever;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static com.example.mike.mp3player.commons.MediaItemType.SONG;
import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.spy;

@RunWith(RobolectricTestRunner.class)
public class SongsFromFolderRetrieverTest extends ContentResolverRetrieverTestBase<SongsFromFolderRetriever> {

    @Before
    public void setup() {
        super.setup();
        this.retriever = spy(new SongsFromFolderRetriever(contentResolver, resultsParser, null, handler));
    }

    @Test
    @Override
    public void testGetMediaType() {
        assertEquals(SONG, retriever.getType());
    }
}

File: service/src/test/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongsFromFolderRetrieverTest.kt
package com.github.goldy1992.mp3player.service.library.content.retriever

import android.os.Looper
import com.github.goldy1992.mp3player.commons.MediaItemBuilder
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.library.content.filter.SongsFromFolderResultsFilter
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser
import com.github.goldy1992.mp3player.service.library.content.request.ContentRequest
import com.github.goldy1992.mp3player.service.library.search.SongDao
import com.nhaarman.mockitokotlin2.*
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows

@RunWith(RobolectricTestRunner::class)
class SongsFromFolderRetrieverTest : ContentResolverRetrieverTestBase<SongsFromFolderRetriever?>() {

    var resultsParser: SongResultsParser = mock<SongResultsParser>()

    var songDao: SongDao = mock<SongDao>()

    @Before
    fun setup() {
        retriever = spy(SongsFromFolderRetriever(contentResolver!!, resultsParser!!, handler, SongsFromFolderResultsFilter()))
    }

    @Test
    fun testGetChildren() {
        contentRequest = ContentRequest("x", "y", "z")
        val id = "xyz"
        val title = "title"
        val mediaItem = MediaItemBuilder(id)
                .setTitle(title)
                .build()
        expectedResult.add(mediaItem)
        whenever(contentResolver!!.query(any(), any(), any(), any(), eq(null))).thenReturn(cursor)
        whenever(resultsParser!!.create(cursor!!, contentRequest!!.mediaIdPrefix!!)).thenReturn(expectedResult)
        val result = retriever!!.getChildren(contentRequest!!)
        // call remaining looper messages
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        // assert results are the expected ones
        Assert.assertEquals(expectedResult, result)
    }

    @Test
    override fun testGetMediaType() {
        Assert.assertEquals(MediaItemType.SONG, retriever!!.type)
    }
}